### PR TITLE
Restructure AES directory and add CPU and GPU code

### PR DESCRIPTION
**Pull request by mhgholamrezaei**
I have added the CPU and GPU codes according to the Vector Addition directory structure.

### CODE REVIEW

**Comment by fasiddique** [comment]
Is this ready to review?

**Comment by mhgholamrezaei** [comment]
Yes. I did not change the PIM code yet. I have added baselines.

File: PIMbench/cpp-aes/README.md [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+```
+
+### Options
+
+- `-l`: Input size (default=65536 bytes)
+- `-k`: Key file containing two vectors (default=generates key with random numbers)
+- `-i`: Input file containing two vectors (default=generates input with random numbers)
+- `-c`: Cipher file containing two vectors (default=./cipher.txt)
+- `-o`: Output file containing two vectors (default=./output.txt)
+- `-v`: (true/false) Validates if the input file and output file are the same (default=false)
```
Comment (fasiddique) [inline PIMbench/cpp-aes/README.md:94]: This is great!

File: PIMbench/cpp-aes/baselines/CPU/aes.cpp [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+    bool generateRandomKey = false;
+    if (params.keyFile == NULL) {
+        generateRandomKey = true;
+        printf("INFO: Key file is not specifed. Random key will be used.\n");
+    }
+    if (!generateRandomKey) {
+        file = fopen(params.keyFile, "r");
+        if (file == NULL) {
+            printf("ERROR: Error opening key file %s\n", params.keyFile);
+        }
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/CPU/aes.cpp:173]:
```suggestion
    if (params.keyFile == NULL) {
        printf("INFO: Key file is not specifed. Random key will be used.\n");
    } else {
        file = fopen(params.keyFile, "r");
        if (file == NULL) {
            printf("ERROR: Error opening key file %s\n", params.keyFile);
        }
```

File: PIMbench/cpp-aes/baselines/CPU/Makefile [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Compiler
+CXX := g++
+
+# Compiler flags
+CXXFLAGS := -Wall -std=c++11 -O3 -fopenmp -lssl -lcrypto
+
+# Output directory
+BUILD_DIR := ./build
+
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/CPU/Makefile:9]:
```suggestion
```

File: PIMbench/cpp-aes/baselines/CPU/Makefile [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# Executable name
+TARGET := $(BUILD_DIR)/aes.out
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/CPU/Makefile:11]:
```suggestion
TARGET := aes.out
```

File: PIMbench/cpp-aes/baselines/CPU/aes.cpp [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    // Read the key from the key file.
+    if (!generateRandomKey && fread(key, 1, 32, file) != 32) {
+        printf("ERROR: The key length in %s is not 32 characters\n", params.keyFile);
+        fclose(file);
+        return EXIT_FAILURE;
+    } else {
+        for (unsigned int i = 0; i < 32; ++i) {
+            key[i] = rand() && 0xff;
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/CPU/aes.cpp:183]: just to make sure, this code intend to store either 0 or 1?

File: PIMbench/cpp-aes/baselines/CPU/aes.cpp [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    // Verify that there are no extra characters.
+    if (!generateRandomKey) {
+        char extra;
+        if (fread(&extra, 1, 1, file) != 0) {
+            printf("ERROR: The key length in %s is more than 32 characters\n", params.keyFile);
+            fclose(file);
+            return EXIT_FAILURE;
+        }
+        fclose(file);
+    }
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/CPU/aes.cpp:196]: This can also be merged.

File: PIMbench/cpp-aes/baselines/CPU/aes.cpp [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        free(plaintext);
+        return EXIT_FAILURE;
+    }
+    if (!generateRandomInput) { 
+        fclose(file);
+    } else {
+        for (unsigned int i = 0; i < params.inputSize; ++i) {
+            plaintext[i] = rand() && 0xff;
+        }
+    }
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/CPU/aes.cpp:240]:
```suggestion
    // Open and read the input file.
    if (params.inputFile == NULL) {
        numbytes = params.inputSize;
        printf("INFO: Input file is not specifed. Random input will be used.\n");
    } else {
        file = fopen(params.inputFile, "r");
        if (file == NULL) {
            printf("ERROR: Error opening input file %s\n", params.inputFile);
        } 
        fseek(file, 0L, SEEK_END);
        numbytes = ftell(file);
        fseek(file, 0L, SEEK_SET);
    }

    // Allocate memory for the file content.
    plaintext = (uint8_t*)malloc((numbytes + 100) * sizeof(uint8_t));
    ciphertext = (uint8_t*)malloc((numbytes + 100) * sizeof(uint8_t));
    decryptedtext = (uint8_t*)malloc((numbytes + 100) * sizeof(uint8_t));
    if (plaintext == NULL || ciphertext == NULL || decryptedtext == NULL) {
        printf("ERROR: Memory allocation error\n");
        fclose(file);
        return EXIT_FAILURE;
    }

    // Read the file into the buffer.
    if (params.inputFile != NULL) {
       if ( fread(plaintext, 1, numbytes, file) != numbytes) {
        printf("ERROR: Unable to read all bytes from file %s\n", params.inputFile);
        fclose(file);
        free(plaintext);
        free(ciphertext);
        free(decryptedtext);
        return EXIT_FAILURE;
        }
        fclose(file);
    } else {
        for (unsigned int i = 0; i < params.inputSize; ++i) {
            plaintext[i] = rand() && 0xff;
        }
    }
```
Please test this.

**Comment by mhgholamrezaei** [comment]
@fasiddique 
Thanks for the code review. I addressed your comments in the recent commit.

File: PIMbench/cpp-aes/baselines/CPU/aes.cpp [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+        }
+        fclose(file);
+    }
+
+    // Allocate memory for the file content.
+    numbytes = params.inputSize;
+    plaintext = (uint8_t*)malloc((numbytes + 100) * sizeof(uint8_t));
+    ciphertext = (uint8_t*)malloc((numbytes + 100) * sizeof(uint8_t));
+    decryptedtext = (uint8_t*)malloc((numbytes + 100) * sizeof(uint8_t));
+    if (plaintext == NULL) {
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/CPU/aes.cpp:194]: shouldn't be checked for `ciphertext` and `decryptedtext` also.

File: PIMbench/cpp-aes/baselines/CPU/aes.cpp [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+        if (file == NULL) {
+            printf("ERROR: Error opening input file %s\n", params.inputFile);
+        } 
+        fseek(file, 0L, SEEK_END);
+        numbytes = ftell(file);
+        fseek(file, 0L, SEEK_SET);
+        if (fread(plaintext, 1, numbytes, file) != numbytes) {
+          printf("ERROR: Unable to read all bytes from file %s\n", params.inputFile);
+          fclose(file);
+          free(plaintext);
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/CPU/aes.cpp:218]: other two should be freed to or else there will be memory leak.

**Comment by fasiddique** [comment]
> @fasiddique Thanks for the code review. I addressed your comments in the recent commit.

Thank you! There are few more issues here.

**Comment by mhgholamrezaei** [comment]
@fasiddique 
The mentioned issues are fixed. Please check it out. Thanks again for the review.

File: PIMbench/cpp-aes/baselines/GPU/aes.cu [hunk 10]
```diff
@@ -0,0 +1,746 @@
...
+  cudaMemcpy(ctx_deckey_d, ctx_deckey, sizeof(ctx_deckey), cudaMemcpyHostToDevice);
+  cudaMemcpy(ctx_key_d, ctx_key, sizeof(ctx_key), cudaMemcpyHostToDevice);
+
+  dim3 dimBlock(ceil((double)numbytes / (double)(THREADS_PER_BLOCK * AES_BLOCK_SIZE)));
+  dim3 dimGrid(THREADS_PER_BLOCK);
+  aes256_decrypt_ecb<<<dimBlock, dimGrid>>>(buf_d, numbytes, ctx_deckey_d, ctx_key_d);
+
+  cudaMemcpy(buf, buf_d, numbytes, cudaMemcpyDeviceToHost);
+  cudaMemcpy(ctx_deckey, ctx_deckey_d, sizeof(ctx_deckey), cudaMemcpyDeviceToHost);
+  cudaMemcpy(ctx_key, ctx_key_d, sizeof(ctx_key), cudaMemcpyDeviceToHost);
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/GPU/aes.cu:489]: Can we please do an error check for cudaMemCpy and cuda kernel calls.

File: PIMbench/cpp-aes/baselines/GPU/aes.cu [hunk 11]
```diff
@@ -0,0 +1,746 @@
...
+  cudaMemcpy(ctx_key_d, ctx_key, sizeof(ctx_key), cudaMemcpyHostToDevice);
+  dim3 dimBlock(ceil((double)numbytes / (double)(THREADS_PER_BLOCK * AES_BLOCK_SIZE)));
+  dim3 dimGrid(THREADS_PER_BLOCK);
+  aes256_encrypt_ecb<<<dimBlock, dimGrid>>>(buf_d, numbytes, ctx_enckey_d, ctx_key_d);
+
+  cudaMemcpy(buf, buf_d, numbytes, cudaMemcpyDeviceToHost);
+  // print(buf);
+  cudaMemcpy(ctx_enckey, ctx_enckey_d, sizeof(ctx_enckey), cudaMemcpyDeviceToHost);
+  cudaMemcpy(ctx_key, ctx_key_d, sizeof(ctx_key), cudaMemcpyDeviceToHost);
+
```
Comment (fasiddique) [inline PIMbench/cpp-aes/baselines/GPU/aes.cu:460]: error check missing.

**Comment by mhgholamrezaei** [comment]
@fasiddique 
Thanks again for catching these errors. I addressed your comments in the most recent commit.

**Review by fasiddique — APPROVED** [review]
Thank you Hosein!

**Pull request merged**